File: src/dataset/javascript/word-search_exercism-javascript-ThomasZumsteg/gemini/GeminiFlash_word-search_v2.js
export default class WordSearch {
  constructor(grid) {
    this.grid = grid;
    this.rows = grid.length;
    this.cols = grid[0] ? grid[0].length : 0;
  }

  find(words) {
    const result = {};
    for (const word of words) {
      result[word] = undefined;
    }

    for (let r = 0; r < this.rows; r++) {
      for (let c = 0; c < this.cols; c++) {
        for (const word of words) {
          if (result[word] === undefined) { // Only search if not already found
            const end = this.search(word, r, c);
            if (end) {
              result[word] = { "start": [r + 1, c + 1], "end": end };
            }
          }
        }
      }
    }
    return result;
  }

  search(word, row, col) {
    const diffs = [
      { "row": 1, "col": 1 },
      { "row": 1, "col": 0 },
      { "row": 1, "col": -1 },
      { "row": 0, "col": 1 },
      { "row": 0, "col": -1 },
      { "row": -1, "col": 1 },
      { "row": -1, "col": 0 },
      { "row": -1, "col": -1 },
    ];

    for (const diff of diffs) {
      let valid = true;
      for (let pos = 0; pos < word.length; pos++) {
        const newRow = diff.row * pos + row;
        const newCol = diff.col * pos + col;

        if (newRow < 0 || newRow >= this.rows || newCol < 0 || newCol >= this.cols || this.grid[newRow][newCol] !== word[pos]) {
          valid = false;
          break;
        }
      }

      if (valid) {
        return [diff.row * (word.length - 1) + row + 1, diff.col * (word.length - 1) + col + 1];
      }
    }

    return undefined;
  }
}

File: src/dataset/go/word_search_Exercism_go_rootulp/gemini/GeminiFlash_word_search_v2.go
package wordsearch

import (
	"fmt"
)

const (
	NORTH      = "NORTH"
	EAST       = "EAST"
	SOUTH      = "SOUTH"
	WEST       = "WEST"
	NORTH_EAST = "NORTH_EAST"
	NORTH_WEST = "NORTH_WEST"
	SOUTH_EAST = "SOUTH_EAST"
	SOUTH_WEST = "SOUTH_WEST"
)

var DIRECTIONS = []string{NORTH, EAST, SOUTH, WEST, NORTH_EAST, NORTH_WEST, SOUTH_EAST, SOUTH_WEST}
var NOT_FOUND_POINT = [2]int{-1, -1}
var NOT_FOUND_LINE = [2][2]int{NOT_FOUND_POINT, NOT_FOUND_POINT}

func Solve(words []string, puzzle []string) (map[string][2][2]int, error) {
	grid := NewGrid(puzzle)
	result := make(map[string][2][2]int, len(words))
	var err error
	for _, word := range words {
		loc, found := grid.Search(word)
		result[word] = loc
		if !found && err == nil {
			err = fmt.Errorf("word %s not found", word)
		}
	}

	return result, err
}

type Grid struct {
	grid [][]rune
	rows int
	cols int
}

func NewGrid(puzzle []string) *Grid {
	rows := len(puzzle)
	cols := 0
	if rows > 0 {
		cols = len(puzzle[0])
	}
	grid := make([][]rune, rows)
	for i, row := range puzzle {
		grid[i] = []rune(row)
	}
	return &Grid{grid: grid, rows: rows, cols: cols}
}

func (g Grid) Search(word string) ([2][2]int, bool) {
	wordLen := len(word)
	if wordLen == 0 {
		return NOT_FOUND_LINE, false
	}

	for row_i := 0; row_i < g.rows; row_i++ {
		for col_i := 0; col_i < g.cols; col_i++ {
			for _, direction := range DIRECTIONS {
				if g.isMatch(word, row_i, col_i, direction) {
					startLocation := [2]int{col_i, row_i}
					endLocation := g.endLocation(startLocation, direction, wordLen)
					return [2][2]int{startLocation, endLocation}, true
				}
			}
		}
	}
	return NOT_FOUND_LINE, false
}

func (g Grid) SafeGet(row int, col int) (rune, bool) {
	if row < 0 || row >= g.rows || col < 0 || col >= g.cols {
		return ' ', false
	}
	return g.grid[row][col], true
}

func (g Grid) isMatch(word string, row int, col int, direction string) bool {
	for i := 0; i < len(word); i++ {
		var r, c int
		switch direction {
		case NORTH:
			r, c = row-i, col
		case EAST:
			r, c = row, col+i
		case SOUTH:
			r, c = row+i, col
		case WEST:
			r, c = row, col-i
		case NORTH_EAST:
			r, c = row-i, col+i
		case NORTH_WEST:
			r, c = row-i, col-i
		case SOUTH_EAST:
			r, c = row+i, col+i
		case SOUTH_WEST:
			r, c = row+i, col-i
		}

		char, ok := g.SafeGet(r, c)
		if !ok || char != rune(word[i]) {
			return false
		}
	}
	return true
}

func (g Grid) endLocation(startLoc [2]int, direction string, wordLen int) (endLoc [2]int) {
	col, row := startLoc[0], startLoc[1]
	delta := wordLen - 1
	switch direction {
	case NORTH:
		return [2]int{col, row - delta}
	case EAST:
		return [2]int{col + delta, row}
	case SOUTH:
		return [2]int{col, row + delta}
	case WEST:
		return [2]int{col - delta, row}
	case NORTH_EAST:
		return [2]int{col + delta, row - delta}
	case NORTH_WEST:
		return [2]int{col - delta, row - delta}
	case SOUTH_EAST:
		return [2]int{col + delta, row + delta}
	case SOUTH_WEST:
		return [2]int{col - delta, row + delta}
	default:
		panic(fmt.Sprintf("invalid direction %v", direction))
	}
}